webapp/store/models.py

```diff
@@ -151,278 +151,294 @@ def _parse_charm_or_bundle(entity_data, include_files=False):
-        "term_ids": _parse_term_ids(meta.get("terms")),
-        "url": ref.jujucharms_id(),
-    }
-
-
-def _parse_term_ids(term_ids):
-    """Extract the term names and revisions.
-        :param term_ids: A list of term ids.
-        :returns: a collection of term ids, names and revisions.
+        return Bundle(entity_data, include_files)
+
+
+class Entity:
+    """Base class for charms and bundles. Contains the shared attributes and
+        methods.
+        Classes that extend Entity will need to provide the following methods:
+        - _get_metadata
+        - _get_metadata
+        - _get_display_name
+        - _get_tags
+        - _get_series
     """
-    if term_ids is None:
-        return None
-    terms = []
-    for term in term_ids:
-        parts = term.split("/")
-        terms.append(
-            {
-                "id": term,
-                "name": parts[0],
-                "revision": int(parts[1]) if len(parts) == 2 else None,
+
+    def __init__(self, entity_data, include_files=False):
+        self.id = entity_data["Id"]
+        self._ref = references.Reference.from_string(self.id)
+        self._meta = entity_data.get("Meta", None)
+        self._metadata = self._get_metadata()
+        description = self._metadata.get("Description")
+        (
+            _,
+            supported,
+            supported_price,
+            supported_description,
+        ) = self._extract_from_extrainfo()
+        bugs_url, homepage = self._extract_from_commoninfo()
+        self.files = None
+        self.readme = None
+        if include_files:
+            self.files = self._get_entity_files(self._meta.get("manifest"))
+            self.readme = self._render_markdown(
+                cs.entity_readme_content(self.id)
+            )
+        self.archive_url = cs.archive_url(self._ref)
+        self.bugs_url = bugs_url
+        self.card_id = self._ref.path()
+        self.channels = self._meta.get("published", {}).get("Info")
+        # Some entities don't have descriptions, so first check that the
+        # description exists before trying to render the Markdown.
+        self.description = (
+            self._render_markdown(description) if description else None
+        )
+        self.display_name = self._get_display_name()
+        self.homepage = homepage
+        self.latest_revision = self._get_latest_revision()
+        self.owner = self._meta.get("owner", {}).get("User")
+        self.promulgated = self._meta.get("promulgated", {}).get("Promulgated")
+        self.revision_number = self._ref.revision
+        self.series = self._get_series()
+        self.supported = supported
+        self.supported_price = supported_price
+        self.supported_description = (
+            supported_description
+            and self._render_markdown(supported_description)
+        )
+        self.tags = self._get_tags()
+        self.url = self._ref.jujucharms_id()
+
+    def _render_markdown(self, content):
```

Comment:
These classes methods are all existing methods that have been moved into the classes (with the occasional self... change and doc strings added).